src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -0,0 +1,327 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers.FrameworkEventSource;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring
+{
+    public enum PipeMode
+    {
+        Logs = 1,
+        Metrics,
+    }
+
+    public class DiagnosticsEventPipeProcessor : IAsyncDisposable
+    {
+        private readonly IServiceProvider _services;
+        private readonly Microsoft.Extensions.Logging.ILogger<DiagnosticsMonitor> _logger;
+        private readonly IEnumerable<IMetricsLogger> _metricLoggers;
+        private readonly PipeMode _mode;
+
+        //These values don't change so we compute them only once
+        private readonly List<string> _dimValues;
+
+        public const string NamespaceName = "Namespace";
+        public const string NodeName = "Node";
+        private static readonly List<string> DimNames = new List<string> { NamespaceName, NodeName };
+
+        public DiagnosticsEventPipeProcessor(IServiceProvider services, PipeMode mode)
```

Comment:
Passing around IServiceProvider is a code smell. Why isn't this class in the container?

Reply:
See my [comment](https://github.com/dotnet/diagnostics/pull/1073/files/6c016902b9bf2a194334658710baab5d005fb194#r414920069)

1. I have the initial service provider.
2. During a logs request, I want to temporarily add additional services (ILogger) that will handle consumption of log requests.
My solution:
Create a new service provider that adds the desired ILogger service.
Create a hierarchical service provider so that previously registered services are also available.
As noted in my comments, I don't really like this solution because the child services cannot actually rely on the parent service provider.

Other things I tried:
- I tried registering my ILogger implementation at startup time (and relied on IHttpContextAccessor to get the writing stream). The problem is that my ILogger will now also be consumed by other parts of asp.net that want to use ILogger services. 
- There are potential timing issues when calling IHttpContextAccessor at the wrong time; the logger service may not always be valid.

Reply:
Seems like the design shouldn’t be around adding new services per request dynamically. Can you restructure so each component minimally gets the required dependencies? A design that leads to some bring like this usually indicates a bigger problem

Reply:
> Can you restructure so each component minimally gets the required dependencies? 
I can do this by passing the required services directly into the class. This will remove the IServiceProvider parameter and the AggregatedServices class. It will also line up better with Noah's proposed design for this Pipeline
>Seems like the design shouldn’t be around adding new services per request dynamically.
This is doable but will still require two service containers. I cannot add my ILogger to the primary service container.

Reply:
@davidfowl 
Here's the updates I made:
- AggregateServices is removed
- IServiceProvider is no longer being passed around
- ILogger services are only constructed once instead of per request

As discussed, there is still a secondary service container for ILogger services designed to be consumed by our pipeline, as opposed to the REST api.

Reply:
Will take a deeper look today.